src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/RewindDataIndexStore.cs

```diff
@@ -0,0 +1,93 @@
+﻿using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using NBitcoin;
+using Stratis.Bitcoin.Features.Consensus.CoinViews;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Features.Consensus.ProvenBlockHeaders
+{
+    /// <inheritdoc />
+    public class RewindDataIndexStore : IRewindDataIndexStore
+    {
+        /// <summary>
+        /// The rewind data index repository that is used to store and retrieve data from the disk.
+        /// </summary>
+        private readonly IRewindDataIndexRepository rewindDataIndexRepository;
+
+        /// <summary>
+        /// Internal cache for rewind data index. Key is a TxId + N (N is an index of output in a transaction)
+        /// and value is a rewind data index.
+        /// </summary>
+        private readonly ConcurrentDictionary<string, int> items;
+
+        /// <summary>
+        /// Number of items to keep in cache after the flush.
+        /// </summary>
+        private const int NumberOfItemsToKeep = 500;
+
+        /// <summary>
+        /// Performance counter to measure performance of the save and get operations.
+        /// </summary>
+        private readonly BackendPerformanceCounter performanceCounter;
+
+        public RewindDataIndexStore(
+            IDateTimeProvider dateTimeProvider,
+            IRewindDataIndexRepository rewindDataIndexRepository)
+        {
+            Guard.NotNull(dateTimeProvider, nameof(dateTimeProvider));
+            Guard.NotNull(rewindDataIndexRepository, nameof(rewindDataIndexRepository));
+
+            this.items = new ConcurrentDictionary<string, int>();
+            this.rewindDataIndexRepository = rewindDataIndexRepository;
+
+            this.performanceCounter = new BackendPerformanceCounter(dateTimeProvider);
+        }
+
+        /// <inheritdoc />
+        public async Task SaveAsync(Dictionary<string, int> indexData)
+        {
+            using (new StopwatchDisposable(o => this.performanceCounter.AddInsertTime(o)))
+            {
+                foreach (KeyValuePair<string, int> indexRecord in indexData)
+                {
+                    this.items[indexRecord.Key] = indexRecord.Value;
+                }
+
+                // Save the items to disk.
+                await this.rewindDataIndexRepository.PutAsync(indexData).ConfigureAwait(false);
```

Comment:
FYI I don think we need to push to disk on save only on flush, @MithrilMan @fenix2222  please fix this in the other PR